pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,384 @@
+
+// processNextWorkItem waits for the queue to have a key available and then processes it.
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	raw, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := raw.(string)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	if err := c.process(ctx, key); err != nil {
+		utilruntime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", controllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	}
+
+	c.queue.Forget(key)
+
+	return true
+}
+
+// process processes a single key from the queue.
+func (c *Controller) process(ctx context.Context, key string) error {
+	// e.g. root:org<separator>ws
+	parent, name := clusters.SplitClusterAwareKey(key)
+
+	// turn it into root:org:ws
+	clusterName := parent.Join(name)
+
+	_, err := c.getClusterWorkspace(key)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			klog.V(2).InfoS("ClusterWorkspace not found - stopping quota controller for it (if needed)", "clusterName", clusterName)
+
+			c.lock.Lock()
+			cancel, ok := c.cancelFuncs[clusterName]
+			if ok {
+				cancel()
+				delete(c.cancelFuncs, clusterName)
+			}
+			c.lock.Unlock()
+
+			c.dynamicDiscoverySharedInformerFactory.Unsubscribe("quota-" + clusterName.String())
+
+			return nil
+		}
+
+		return err
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	_, found := c.cancelFuncs[clusterName]
+	if found {
+		klog.V(4).InfoS("Quota controller already exists", "clusterName", clusterName)
+		return nil
+	}
+
+	klog.V(2).InfoS("Starting quota controller", "clusterName", clusterName)
+
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelFuncs[clusterName] = cancel
+
+	if err := c.startQuotaForClusterWorkspace(ctx, clusterName); err != nil {
+		return fmt.Errorf("error starting quota controller for cluster %q: %w", clusterName, err)
+	}
+
+	return nil
+}
+
+func (c *Controller) startQuotaForClusterWorkspace(ctx context.Context, clusterName logicalcluster.Name) error {
+	resourceQuotaControllerClient := c.kubeClusterClient.Cluster(clusterName)
+
+	// TODO(ncdc): find a way to support the default configuration. For now, don't use it, because it is difficult
+	// to get support for the special evaluators for pods/services/pvcs.
+	// listerFuncForResource := generic.ListerFuncForResourceFunc(scopedInformerFactory.ForResource)
+	// quotaConfiguration := install.NewQuotaConfigurationForControllers(listerFuncForResource)
+	quotaConfiguration := generic.NewConfiguration(nil, install.DefaultIgnoredResources())
+
+	resourceQuotaControllerOptions := &resourcequota.ControllerOptions{
+		QuotaClient: resourceQuotaControllerClient.CoreV1(),
+		ResourceQuotaInformer: &SingleClusterResourceQuotaInformer{
+			clusterName:     clusterName,
+			delegate:        c.resourceQuotaInformer,
+			registerHandler: c.registerResourceQuotaEventHandlerLocked,
```

Comment:
have a hard time to see why it is guaranteed here that the lock is taken. Leaking a `*Locked` func into another object is suspicious.

Reply:
Yeah, I know what you mean. Here is the flow:

- kcp kubequota Controller.process
- it grabs the lock
- calls startQuotaForClusterWorkspace
- calls resourcequota.NewController(resourceQuotaControllerOptions)
- calls options.ResourceQuotaInformer.Informer().AddEventHandlerWithResyncPeriod(
- this is to our SingleClusterResourceQuotaInformer/delegatingInformer
- calls c.registerResourceQuotaEventHandlerLocked

But, let me make add a "level above" and do something similar to `scopingGenericSharedInformerFactory`. Stay tuned. 😄

Reply:
👀